File: 2020/Day08/Python/HandheldHalting-Part1.py
import sys
import collections
import itertools
import numpy as np

acc = 0
instruction = 0
code = []
seen = []

for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    op, arg = line.split()
    code.append((op, int(arg)))

while instruction not in seen:
    seen.append(instruction)
    op, arg = code[instruction]
    if op == "nop":
        instruction += 1
    elif op == "acc":
        acc += arg
        instruction += 1
    elif op == "jmp":
        instruction += arg
    
print(acc)

File: 2020/Day08/Haskell/HandheldHalting-Part1.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

strip:: (Eq a) => [a] -> a -> [a]
strip [] _ = []
strip (x:xs) a
    | x == a = strip xs a
    | otherwise = x : strip xs a

runCode:: Int -> Int -> [Int] -> [([Char], Int)] -> Int
runCode acc ins seen code 
    | ins `elem` seen = acc
    | op == "nop" = runCode acc (ins+1) (ins:seen) code
    | op == "jmp" = runCode acc (ins+arg) (ins:seen) code
    | op == "acc" = runCode (acc+arg) (ins+1) (ins:seen) code
    where (op, arg) = code!!ins 

parseInstruction:: [Char] -> ([Char], Int)
parseInstruction ins = (op, read arg::Int)
    where (op:arg:_) = split ' ' (strip ins '+')

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    print (runCode 0 0 [] (map parseIns rows))
    hClose handle